source/common/common/perf_annotation.cc

```diff
@@ -0,0 +1,149 @@
+  std::string out;
+#if PERF_THREAD_SAFE
+  std::unique_lock<std::mutex> lock(context->mutex_);
+#endif
+
+  // The map is from category/description -> [duration, time]. Reverse-sort by duration.
+  std::vector<const DurationStatsMap::value_type*> sorted_values;
+  sorted_values.reserve(context->duration_stats_map_.size());
+  for (const auto& iter : context->duration_stats_map_) {
+    sorted_values.push_back(&iter);
+  }
+  std::sort(
+      sorted_values.begin(), sorted_values.end(),
+      [](const DurationStatsMap::value_type* a, const DurationStatsMap::value_type* b) -> bool {
+        const DurationStats& a_stats = a->second;
+        const DurationStats& b_stats = b->second;
+        return a_stats.total_ > b_stats.total_;
+      });
+
+  // Organize the report so it lines up in columns. Note that the widest duration comes first,
+  // though that may not be descending order of calls or per_call time, so we need two passes
+  // to compute column widths. First collect the column headers and their widths.
+  //
+  // TODO(jmarantz): add more stats, e.g. std deviation, median, min, max.
+  static const char* headers[] = {"Duration(us)", "# Calls", "Mean(ns)", "StdDev(ns)",
+                                  "Min(ns)",      "Max(ns)", "Category", "Description"};
+  constexpr int num_columns = ARRAY_SIZE(headers);
+  size_t widths[num_columns];
+  std::vector<std::string> columns[num_columns];
+  for (size_t i = 0; i < num_columns; ++i) {
+    std::string column(headers[i]);
+    widths[i] = column.size();
+    columns[i].emplace_back(column);
+  }
+
+  // Compute all the column strings and their max widths.
+  for (const auto& p : sorted_values) {
+    const DurationStats& stats = p->second;
+    auto microseconds_string = [](std::chrono::nanoseconds ns) -> std::string {
+      return std::to_string(std::chrono::duration_cast<std::chrono::microseconds>(ns).count());
+    };
+    auto nanoseconds_string = [](std::chrono::nanoseconds ns) -> std::string {
+      return std::to_string(std::chrono::duration_cast<std::chrono::nanoseconds>(ns).count());
+    };
+    columns[0].push_back(microseconds_string(stats.total_));
+    uint64_t count = stats.stddev_.count();
+    columns[1].push_back(std::to_string(count));
+    columns[2].push_back(
+        (count == 0)
+            ? "NaN"
+            : std::to_string(
+                  std::chrono::duration_cast<std::chrono::nanoseconds>(stats.total_).count() /
+                  count));
+    columns[3].push_back(fmt::format("{}", stats.stddev_.computeStandardDeviation()));
+    columns[4].push_back(nanoseconds_string(stats.min_));
+    columns[5].push_back(nanoseconds_string(stats.max_));
+    const CategoryDescription& category_description = p->first;
+    columns[6].push_back(category_description.first);
+    columns[7].push_back(category_description.second);
+    for (size_t i = 0; i < num_columns; ++i) {
+      widths[i] = std::max(widths[i], columns[i].back().size());
+    }
```

Comment:
FWIW, I'm thinking just dumping to HTML might be even easier, since then there is no manual layout engine work to be done, but this is fine.

Reply:
I would rather generate columnar text for easy fast consumption while developing.  I think an HTML table generator would also be very useful if/when this is hooked to the admin console, so I added a TODO for that.